File: packages/kits/GeniusLocusKit/Tests/GLKMigrationV1_3ToV1_4Tests/IndexCompositionSettingMigrationTests.swift
#if GLK_MIGRATION_V1_3_TO_V1_4

// IndexCompositionSettingMigrationTests.swift
//
// Verifies that GLKMigrationCatalog.prepare runs the 1.3→1.4 capsule, which
// stores the index composition setting (manifest key
// index_composition_policy) on populated estates written before the setting
// existed and stamps the estate format v1_4.
//
// Tests:
//   1. v1_3-stamped estate without the setting: after prepare the setting
//      reads `.current` (no MOOT_INDEX_COMPOSITION in the environment) and
//      the estate is stamped v1_4.
//   2. Idempotence: a second prepare is a no-op that leaves the stamp at
//      v1_4 and the setting untouched.
//   3. A stored setting survives the capsule: an estate that already carries
//      cell B keeps cell B.
//   4. MOOT_INDEX_COMPOSITION set to a valid id at upgrade time seeds that
//      id; an invalid value seeds `.current`.
//   5. Fresh estate (nil stamp): prepare stamps v1_4 and seeds the setting
//      without running any capsule.
//   6. v1_0-stamped estate: the full chain (v1_0→v1_1→v1_2→v1_3→v1_4) ends
//      at v1_4 with the setting stored.

import CorpusKit
import Foundation
import LocusKit
import PersistenceKit
import PersistenceKitInMemory
import Testing
import GeniusLocusKitMigrations

@testable import GeniusLocusKit
@testable import GLKMigrationV1_3ToV1_4

private let testOwner = OwnerCredentials(ownerIdentifier: "test-owner-mig14")
private let testNow = Date(timeIntervalSince1970: 1_756_000_000)

/// An in-memory estate stamped at `stamp` (or unstamped when nil) with no
/// index composition setting, opened in a fresh GeniusLocusKit instance.
private func makeEstate(
    stampedAt stamp: EstateFormatVersion?
) async throws -> (kit: GeniusLocusKit, handle: EstateHandle, storage: InMemoryStorage) {
    let storage = InMemoryStorage(
        configuration: EstateConfiguration(estateID: UUID(), backend: .inMemory))
    _ = try await LocusKit.Estate.create(storage: storage, owner: testOwner)
    if let stamp {
        try await EstateFormatStore(storage: storage).stamp(stamp, now: testNow)
    }
    let kit = GeniusLocusKit()
    let handle = try await kit.open(
        storage: storage,
        owner: testOwner,
        identityKeyStore: InMemoryEstateIdentityKeyStore()
    )
    return (kit, handle, storage)
}

/// Run `body` with MOOT_INDEX_COMPOSITION set to `value` (or removed when
/// nil) and restore the previous value afterwards. The suite is serialized,
/// so no other test in it observes the window.
private func withCreationSeed<T>(
    _ value: String?, _ body: () async throws -> T
) async throws -> T {
    let key = GeniusLocusKit.indexCompositionPolicyEnvironmentKey
    let previous = ProcessInfo.processInfo.environment[key]
    if let value { setenv(key, value, 1) } else { unsetenv(key) }
    defer {
        if let previous { setenv(key, previous, 1) } else { unsetenv(key) }
    }
    return try await body()
}

@Suite("IndexCompositionSettingMigration", .serialized)
struct IndexCompositionSettingMigrationTests {

    // MARK: §1 Core: v1_3 estate gains the stored setting and stamps v1_4

    @Test
    func v1_3EstateGainsStoredSetting() async throws {
        try await withCreationSeed(nil) {
            let (kit, handle, storage) = try await makeEstate(stampedAt: .v1_3)
            #expect(try await kit.storedIndexCompositionPolicy(for: handle) == nil)

            let prep = try await GLKMigrationCatalog.prepare(kit: kit, handle: handle, now: testNow)
            #expect(prep.format == .v1_4)
            #expect(prep.format == .current)
            #expect(prep.migrated == false)
            #expect(try await EstateFormatStore(storage: storage).readIfPresent() == .v1_4)
            #expect(try await kit.storedIndexCompositionPolicy(for: handle) == .current)
            // The row is the policy id, verbatim.
            let estate = try await kit.estate(for: handle)
            #expect(try await estate.meta(key: GeniusLocusKit.indexCompositionPolicyMetaKey)
                == IndexCompositionPolicy.current.id)
        }
    }

    // MARK: §2 Idempotence

    @Test
    func prepareTwiceIsNoOp() async throws {
        try await withCreationSeed(nil) {
            let (kit, handle, storage) = try await makeEstate(stampedAt: .v1_3)
            let first = try await GLKMigrationCatalog.prepare(kit: kit, handle: handle, now: testNow)
            #expect(first.format == .v1_4)
            let second = try await GLKMigrationCatalog.prepare(kit: kit, handle: handle, now: testNow)
            #expect(second.format == .v1_4)
            #expect(second.migrated == false)
            #expect(try await EstateFormatStore(storage: storage).readIfPresent() == .v1_4)
            #expect(try await kit.storedIndexCompositionPolicy(for: handle) == .current)
        }
    }

    // MARK: §3 A stored setting survives the capsule

    @Test
    func storedSettingIsLeftUntouched() async throws {
        try await withCreationSeed(IndexCompositionPolicy.lexicalBaseline.id) {
            let (kit, handle, _) = try await makeEstate(stampedAt: .v1_3)
            try await kit.setIndexCompositionPolicy(.lexicalAdornments, for: handle)
            // Running the capsule directly (not through prepare) proves the
            // seed step itself leaves a stored value alone, even with a
            // different creation seed in the environment.
            try await kit.runIndexCompositionSettingMigration(handle: handle, now: testNow)
            #expect(try await kit.storedIndexCompositionPolicy(for: handle) == .lexicalAdornments)
        }
    }

    // MARK: §4 The creation seed at upgrade time

    @Test
    func environmentSeedsTheSettingAtUpgradeTime() async throws {
        try await withCreationSeed(IndexCompositionPolicy.bothAdornments.id) {
            let (kit, handle, _) = try await makeEstate(stampedAt: .v1_3)
            _ = try await GLKMigrationCatalog.prepare(kit: kit, handle: handle, now: testNow)
            #expect(try await kit.storedIndexCompositionPolicy(for: handle) == .bothAdornments)
        }
        try await withCreationSeed("lex=nonsense;dense=distilled") {
            let (kit, handle, _) = try await makeEstate(stampedAt: .v1_3)
            _ = try await GLKMigrationCatalog.prepare(kit: kit, handle: handle, now: testNow)
            #expect(try await kit.storedIndexCompositionPolicy(for: handle) == .current)
        }
    }

    // MARK: §5 Fresh estate (nil stamp)

    @Test
    func freshEstateStampsCurrentAndSeedsWithoutCapsule() async throws {
        try await withCreationSeed(nil) {
            let (kit, handle, storage) = try await makeEstate(stampedAt: nil)
            let prep = try await GLKMigrationCatalog.prepare(kit: kit, handle: handle, now: testNow)
            #expect(prep.format == .current)
            #expect(prep.format == .v1_4)
            #expect(prep.migrated == false)
            #expect(prep.migrationState == nil)
            #expect(try await EstateFormatStore(storage: storage).readIfPresent() == .v1_4)
            #expect(try await kit.storedIndexCompositionPolicy(for: handle) == .current)
        }
    }

    // MARK: §6 Full chain from v1_0 ends at v1_4

    #if GLK_MIGRATION_V1_0_TO_V1_1
    @Test
    func v1_0EstateRunsFullChainToCurrent() async throws {
        try await withCreationSeed(nil) {
            let (kit, handle, storage) = try await makeEstate(stampedAt: .v1_0)
            let prep = try await GLKMigrationCatalog.prepare(kit: kit, handle: handle, now: testNow)
            #expect(prep.format == .v1_4)
            #expect(prep.migrated == false)
            #expect(try await EstateFormatStore(storage: storage).readIfPresent() == .v1_4)
            #expect(try await kit.storedIndexCompositionPolicy(for: handle) == .current)
        }
    }
    #endif
}

#endif // GLK_MIGRATION_V1_3_TO_V1_4

File: packages/kits/GeniusLocusKit/rust-migrations/tests/index_composition_setting_migration_tests.rs
//! Verification fixture for the GLK 1.3 → 1.4 index-composition-setting
//! migration. Rust twin of Swift `IndexCompositionSettingMigrationTests.swift`.
//!
//! Tests:
//!   1. v1_3-stamped estate without the setting: after the capsule the
//!      setting reads `current()` (no MOOT_INDEX_COMPOSITION) and the estate
//!      is stamped V1_4.
//!   2. Idempotence: a second run leaves the stamp at V1_4 and the setting
//!      untouched.
//!   3. A stored setting survives the capsule (cell B stays cell B).
//!   4. MOOT_INDEX_COMPOSITION set to a valid id at upgrade time seeds that
//!      id; an invalid value seeds `current()`. Both cases share one test so
//!      the environment mutation never races another test.
//!   5. StorageUnavailable: an unregistered handle returns the error variant.
//!   6. Full chain from v1_0 through the compiled catalog ends at V1_4 with
//!      the setting stored. (Gated on feature = "migration-v1-0-to-v1-1".)

use std::sync::Arc;

use corpus_kit::index_composition_policy::IndexCompositionPolicy;
use genius_locus_kit::estate_format::{EstateFormatStore, EstateFormatVersion};
use genius_locus_kit::EstateCoordinator;
use genius_locus_kit_migrations::{
    IndexCompositionSettingMigrationError, IndexCompositionSettingMigrationExt,
};
use locus_kit::drawer_store::DrawerStore;
use locus_kit::drawer_store_inmemory::InMemoryDrawerStore;
use locus_kit::estate_types::OwnerCredentials;
use persistence_kit::Storage;

const NOW: i64 = 1_756_000_000_000; // millis

/// Open an in-memory estate stamped at `stamp_version` with no index
/// composition setting.
fn make_estate(
    stamp_version: EstateFormatVersion,
) -> (
    EstateCoordinator,
    genius_locus_kit::handle::EstateHandle,
    Arc<dyn Storage>,
) {
    let store = Arc::new(InMemoryDrawerStore::new(NOW, None).expect("InMemoryDrawerStore::new"));
    let storage: Arc<dyn Storage> = store
        .storage()
        .expect("InMemoryDrawerStore exposes storage");
    EstateFormatStore::new(Arc::clone(&storage))
        .stamp(stamp_version, NOW)
        .expect("stamp estate format");
    let mut coord = EstateCoordinator::new();
    let handle = coord
        .open(
            Arc::clone(&store) as Arc<dyn DrawerStore>,
            OwnerCredentials::new("mig14-test-owner"),
            0,
            100,
        )
        .expect("EstateCoordinator::open");
    (coord, handle, storage)
}

fn read_stamp(storage: &Arc<dyn Storage>) -> EstateFormatVersion {
    EstateFormatStore::new(Arc::clone(storage))
        .read_if_present()
        .expect("read format version")
        .expect("version must be set")
}

/// Run `body` with MOOT_INDEX_COMPOSITION set to `value` (removed when None)
/// and restore the previous value afterwards. The process environment is
/// shared by every test thread, so the seed window is serialized through
/// one lock; a poisoned lock (a failed test) is recovered rather than
/// cascading into every later test.
fn with_creation_seed<T>(value: Option<&str>, body: impl FnOnce() -> T) -> T {
    static SEED_LOCK: std::sync::Mutex<()> = std::sync::Mutex::new(());
    let _serialized = SEED_LOCK.lock().unwrap_or_else(|poisoned| poisoned.into_inner());
    let key = EstateCoordinator::INDEX_COMPOSITION_POLICY_ENV_KEY;
    let previous = std::env::var(key).ok();
    match value {
        Some(v) => std::env::set_var(key, v),
        None => std::env::remove_var(key),
    }
    let out = body();
    match previous {
        Some(p) => std::env::set_var(key, p),
        None => std::env::remove_var(key),
    }
    out
}

// ---------------------------------------------------------------------------
// §1 Core: v1_3-stamped estate gains the stored setting and stamps V1_4
// ---------------------------------------------------------------------------

#[test]
fn v1_3_estate_gains_stored_setting_and_stamps_v1_4() {
    with_creation_seed(None, || {
        let (coord, handle, storage) = make_estate(EstateFormatVersion::V1_3);
        assert_eq!(coord.stored_index_composition_policy(&handle).expect("read"), None);

        let policy = coord
            .run_index_composition_setting_migration(&handle, NOW)
            .expect("migration must succeed on v1_3 estate");

        assert_eq!(policy, IndexCompositionPolicy::current());
        assert_eq!(read_stamp(&storage), EstateFormatVersion::V1_4);
        assert_eq!(read_stamp(&storage), EstateFormatVersion::CURRENT);
        assert_eq!(
            coord.stored_index_composition_policy(&handle).expect("read"),
            Some(IndexCompositionPolicy::current())
        );
        // The row is the policy id, verbatim.
        let estate = coord.estate_for(&handle).expect("estate");
        assert_eq!(
            estate
                .meta(EstateCoordinator::index_composition_policy_meta_key())
                .expect("meta"),
            Some(IndexCompositionPolicy::current().id())
        );
    });
}

// ---------------------------------------------------------------------------
// §2 Idempotence
// ---------------------------------------------------------------------------

#[test]
fn capsule_is_idempotent() {
    with_creation_seed(None, || {
        let (coord, handle, storage) = make_estate(EstateFormatVersion::V1_3);
        coord
            .run_index_composition_setting_migration(&handle, NOW)
            .expect("first call must succeed");
        coord
            .run_index_composition_setting_migration(&handle, NOW)
            .expect("second call must be idempotent");
        assert_eq!(read_stamp(&storage), EstateFormatVersion::V1_4);
        assert_eq!(
            coord.stored_index_composition_policy(&handle).expect("read"),
            Some(IndexCompositionPolicy::current())
        );
    });
}

// ---------------------------------------------------------------------------
// §3 A stored setting survives the capsule
// ---------------------------------------------------------------------------

#[test]
fn stored_setting_is_left_untouched() {
    with_creation_seed(Some(&IndexCompositionPolicy::lexical_baseline().id()), || {
        let (coord, handle, _) = make_estate(EstateFormatVersion::V1_3);
        coord
            .set_index_composition_policy(&handle, IndexCompositionPolicy::lexical_adornments())
            .expect("set");
        coord
            .run_index_composition_setting_migration(&handle, NOW)
            .expect("migration");
        assert_eq!(
            coord.stored_index_composition_policy(&handle).expect("read"),
            Some(IndexCompositionPolicy::lexical_adornments())
        );
    });
}

// ---------------------------------------------------------------------------
// §4 The creation seed at upgrade time
// ---------------------------------------------------------------------------

#[test]
fn environment_seeds_the_setting_at_upgrade_time() {
    with_creation_seed(Some(&IndexCompositionPolicy::both_adornments().id()), || {
        let (coord, handle, _) = make_estate(EstateFormatVersion::V1_3);
        coord
            .run_index_composition_setting_migration(&handle, NOW)
            .expect("migration");
        assert_eq!(
            coord.stored_index_composition_policy(&handle).expect("read"),
            Some(IndexCompositionPolicy::both_adornments())
        );
    });
    with_creation_seed(Some("lex=nonsense;dense=distilled"), || {
        let (coord, handle, _) = make_estate(EstateFormatVersion::V1_3);
        coord
            .run_index_composition_setting_migration(&handle, NOW)
            .expect("migration");
        assert_eq!(
            coord.stored_index_composition_policy(&handle).expect("read"),
            Some(IndexCompositionPolicy::current())
        );
    });
}

// ---------------------------------------------------------------------------
// §5 StorageUnavailable
// ---------------------------------------------------------------------------

#[test]
fn unregistered_handle_returns_storage_unavailable() {
    let coord = EstateCoordinator::new();
    let store = InMemoryDrawerStore::new(NOW, None).expect("InMemoryDrawerStore::new");
    let mut staging = EstateCoordinator::new();
    let unregistered_handle = staging
        .open(
            Arc::new(store),
            OwnerCredentials::new("unregistered-owner"),
            0,
            100,
        )
        .expect("open in staging coord");
    let result = coord.run_index_composition_setting_migration(&unregistered_handle, NOW);
    assert!(
        matches!(result, Err(IndexCompositionSettingMigrationError::StorageUnavailable { .. })),
        "expected StorageUnavailable, got {result:?}"
    );
}

// ---------------------------------------------------------------------------
// §6 Full chain from v1_0 (only when migration-floor-1-0 is enabled)
// ---------------------------------------------------------------------------

#[cfg(feature = "migration-v1-0-to-v1-1")]
#[test]
fn v1_0_estate_runs_full_chain_to_v1_4() {
    use corpus_kit_providers::default_ensemble;
    use genius_locus_kit_migrations::{compiled_floor, MigrationChainExt};

    with_creation_seed(None, || {
        let (mut coord, handle, storage) = make_estate(EstateFormatVersion::V1_0);
        assert_eq!(compiled_floor(), Some(EstateFormatVersion::V1_0));
        coord
            .run_migration_chain(&handle, NOW, default_ensemble())
            .expect("full chain must succeed on an empty v1_0 estate");
        assert_eq!(read_stamp(&storage), EstateFormatVersion::V1_4, "full chain must end at V1_4");
        assert_eq!(
            coord.stored_index_composition_policy(&handle).expect("read"),
            Some(IndexCompositionPolicy::current())
        );
    });
}
